python/tests/test_logistic_regression.py

```diff
@@ -68,32 +68,36 @@ def test_toy_example(gpu_number: int) -> None:
         )
         assert lr_model.intercept == pytest.approx(0, abs=1e-6)
 
-        preds_df_local = lr_model.transform(df).collect()
-        preds = [row["prediction"] for row in preds_df_local]
-        assert preds == [1.0, 1.0, 0.0, 0.0]
-        probs = [row["probs"] for row in preds_df_local]
-        assert len(probs) == len(preds)
-        assert [p[1] > 0.5 for p in probs] == [True, True, False, False]
+        def assert_transform(model: LogisticRegressionModel) -> None:
+            preds_df_local = model.transform(df).collect()
+            preds = [row["prediction"] for row in preds_df_local]
+            assert preds == [1.0, 1.0, 0.0, 0.0]
+            probs = [row["probs"] for row in preds_df_local]
+            assert len(probs) == len(preds)
+            assert [p[1] > 0.5 for p in probs] == [True, True, False, False]
+
+        assert_transform(lr_model)
 
         # test with regParam set to 0
         with pytest.warns():
             lr_regParam_zero = LogisticRegression(
                 regParam=0.0,
             )
 
+        lr_regParam_zero.setProbabilityCol(probability_col)
+
         assert lr_regParam_zero.getRegParam() == 0
         assert lr_regParam_zero.cuml_params["C"] == 1.0 / sys.float_info.min
-        model = lr_regParam_zero.fit(df)
-        assert model.coefficients.toArray() == pytest.approx(
-            [-17.21179962158203, 17.220483779907227], abs=1e-6
```

Comment:
Found coefficients are different when running the test case on different GPUs. The values are sensitive to GPU hardware when regParam is set to sys.float_info.min

Reply:
Do we know why?   What we set to smallest float32 value?

Reply:
np.finfo('float32').tiny

Reply:
Observed the same behavior with np.finfo('float32').tiny:
(1) CUDA_VISIBLE_DEVICES=0 pytest -s --runslow tests/test_logistic_regression.py::test_toy_example
	tiny coefficients: [-17.452407836914062,17.457918167114258]                     
	tiny intercept: 0.00544205866754055
2) CUDA_VISIBLE_DEVICES=1 pytest -s --runslow tests/test_logistic_regression.py::test_toy_example
	tiny coefficients: [-16.82124137878418,16.906503677368164]                      
	tiny intercept: 0.03209238499403

GPU 0 is a A5000, GPU 1 is a T400

Reply:
Results of regParam = Sys.float_info.min

(1) CUDA_VISIBLE_DEVICES=0 pytest -s --runslow tests/test_logistic_regression.py::test_toy_example
	coefficients: [-17.452407836914062,17.457918167114258]                          
	intercept: 0.00544205866754055
(2) CUDA_VISIBLE_DEVICES=1 pytest -s --runslow tests/test_logistic_regression.py::test_toy_example
	coefficients: [-16.82124137878418,16.906503677368164]                           
	intercept: 0.03209238499403

It seems np.float('float32').tiny and sys.float_info.min lead to the same converged coefficients on the same GPU.

Reply:
Note the CI GPU presents the same coefficients and intercept as my T400.

Reply:
Any idea at what value they start to diverge?

Reply:
I meant what value for regParam.   How small before the GPUs are inconsistent?

Reply:
Do you mean sweeping regParam? I only tried sys.float_info.min and np.finfo('float32').tiny.

Reply:
Yes.   Or is there a discrepancy for all values for regParam?

Reply:
It seems the smallest value is 1e-4 if the tolerance is 0.001, and is 1e-5 if the tolerance is 0.005.


1e-1
tiny is 0.1
[[-1.8472011  1.8472011]]
[1.9790605e-09]
tiny is 0.1
[[-1.847201  1.847201]]
[-5.2670657e-08]

1e-2
tiny is 0.01
[[-3.411384  3.411398]]
[-2.5473419e-06]
tiny is 0.01
[[-3.4118302  3.4117336]]
[-3.234233e-05]

1e-3
tiny is 0.001
[[-5.2539444  5.2540293]]
[-0.00014487]
tiny is 0.001
[[-5.2539454  5.254031 ]]
[-0.00014471]

1e-4
tiny is 0.0001
[[-7.2327857  7.231805 ]]
[0.00155628]
tiny is 0.0001
[[-7.231882  7.233059]]
[-0.00180929]

1e-5
tiny is 1e-05
[[-9.281207  9.278979]]
[-0.0004311]
tiny is 1e-05
[[-9.2682905  9.269007 ]]
[-0.00121753]

1e-10
tiny is 1e-10
[[-17.33858   17.357773]]
[-0.5866821]
tiny is 1e-10
[[-17.428919  17.273905]]
[-0.04930465]

1e-15
tiny is 1e-15
[[-16.825907  16.75459 ]]
[-0.02975481]
tiny is 1e-15
[[-17.478167  17.32451 ]]
[-0.04882036]

1e-20
tiny is 1e-20
[[-16.825907  16.75459 ]]
[-0.02973408]
tiny is 1e-20
[[-17.478167  17.324514]]
[-0.04881991]

1e-25
tiny is 1e-25
[W] [22:09:16.252830] L-BFGS stopped, because the line search failed to advance (step delta = 0.000000)
[[-16.825907  16.75459 ]]
[-0.02973408]
tiny is 1e-25
[[-17.478167  17.324514]]
[-0.04881991]

1e-30
tiny is 1e-30
[W] [22:09:25.561396] L-BFGS stopped, because the line search failed to advance (step delta = 0.000000)
[[-16.825907  16.75459 ]]
[-0.02973408]
tiny is 1e-30
[W] [22:09:30.109732] L-BFGS stopped, because the line search failed to advance (step delta = 0.000000)
[[-17.478167  17.324514]]
[-0.04881991]

Reply:
Do we want to map regParam 0 to 1e-4 instead of sys.float_info.min? 

I vote for sys.float_info.min as Spark CPU regParam 0 produces coefficients [-17.655434891154886,17.655434889668594]. This is close to sys.float_info.min coefficients.

Reply:
The results suggest a fundamental discrepancy between the two GPUs.   It seems the results are different even for not very small regParam.    GPU 0 stabilizes nicely as it shrinks while GPU 1 at 1e-10 changes direction.    GPU 1 is not really meant for ML/DL, but would be good to understand what is going on here.

Reply:
Not following the point "GPU 1 at 1e-10 changes direction". 

Seem we need to look into C++ implementation to gain more understanding.

Reply:
Replaced sys.float_info.min with np.finfo('float32').tiny to avoid potential type mismatch of float32 and float64.